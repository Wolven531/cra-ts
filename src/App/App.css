@import '../vars.css';

.app {
	background-color: var(--halloween-gray);
	padding: var(--spacing) 0;
	color: var();
}

.content {
	border: 1px solid var(--halloween-orange);
	border-radius: 5px;
	box-shadow: 0 0 5px 5px var(--halloween-black-faded);
	box-sizing: border-box;
	color: var(--halloween-orange);
	display: flex;
	height: var(--content-height);
	margin: var(--spacing) auto;
	padding: var(--spacing);
	width: var(--content-width);
}

footer {
	box-sizing: border-box;
	color: var(--halloween-orange);
	display: flex;
	justify-content: center;
	margin: var(--spacing) auto;
	text-shadow: 1px 1px var(--halloween-black-faded);
	width: var(--content-width);
}

header {
	box-sizing: border-box;
	color: var(--halloween-orange);
	display: flex;
	justify-content: center;
	margin: var(--spacing) auto;
	text-shadow: 1px 1px var(--halloween-black-faded);
	width: var(--content-width);
}
