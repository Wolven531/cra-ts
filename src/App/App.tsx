import React, { FC, useCallback, useEffect, useState } from 'react'
import { useHistory, useParams } from 'react-router-dom'
import { Tabs } from '../Tabs'
import './App.css'

export interface AppProps {
	initialTab?: number
}

export const App: FC<AppProps> = ({ initialTab }) => {
	const history = useHistory()
	const params = useParams<{ tabId?: string }>()

	const [selectedTab, setSelectedTab] = useState(
		initialTab !== undefined ? initialTab : 1
	)

	const handleTabChange = useCallback(
		(tabNum: number) => {
			setSelectedTab(tabNum)
			history.push(`/tabs/${tabNum}`)
		},
		[history, setSelectedTab]
	)

	/**
	 * Runs on component mount
	 */
	useEffect(() => {
		if (params.tabId === undefined) {
			handleTabChange(selectedTab)
			return
		}

		const paramTab = parseInt(params.tabId)
		handleTabChange(paramTab)
	}, [handleTabChange, params.tabId, selectedTab])

	return (
		<article className="app" data-testid="app">
			<header>
				<h3>TS Playground</h3>
			</header>
			<Tabs onTabChange={handleTabChange} selectedTab={selectedTab} />
			<section className="content" data-testid="content">
				{selectedTab === 1 && <div>Content 1</div>}
				{selectedTab === 2 && <div>Content 2</div>}
				{selectedTab === 3 && <div>Content 3</div>}
			</section>
			<footer>&copy; 2021 Anthony Williams</footer>
		</article>
	)
}
