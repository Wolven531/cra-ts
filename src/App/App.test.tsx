import React from 'react'
import { render, RenderResult, waitFor } from '@testing-library/react'
import { App } from './App'
import { MemoryRouter, Route } from 'react-router-dom'

describe('App', () => {
	let comp: RenderResult

	beforeEach(() => {
		comp = render(
			<MemoryRouter initialEntries={['/']} initialIndex={0}>
				<Route path={['/tabs/:tabId', '/tabs', '/']}>
					<App />
				</Route>
			</MemoryRouter>
		)
	})

	it('matches loading snapshot', () => {
		expect(comp.asFragment()).toMatchSnapshot()
	})

	describe('when loaded', () => {
		beforeEach(async () => {
			await waitFor(() => comp.getByTestId('app'))
		})

		it('displays footer text', () => {
			expect(comp.getByText(/Â© 2021 Anthony Williams/i)).toBeInTheDocument()
		})

		it('displays header text', () => {
			expect(comp.getByText(/TS Playground/i)).toBeInTheDocument()
		})

		it('matches snapshot', () => {
			expect(comp.asFragment()).toMatchSnapshot()
		})
	})
})

describe('App w/ initialTab=2 when loaded', () => {
	let comp: RenderResult

	beforeEach(async () => {
		comp = render(
			<MemoryRouter initialEntries={['/tabs/2']} initialIndex={0}>
				<Route path={['/tabs/:tabId', '/tabs', '/']}>
					<App initialTab={2} />
				</Route>
			</MemoryRouter>
		)

		await waitFor(() => comp.getByTestId('app'))
	})

	it('matches snapshot', () => {
		expect(comp.asFragment()).toMatchSnapshot()
	})
})

describe('App w/ initialTab=3 when loaded', () => {
	let comp: RenderResult

	beforeEach(async () => {
		comp = render(
			<MemoryRouter initialEntries={['/tabs/3']} initialIndex={0}>
				<Route path={['/tabs/:tabId', '/tabs', '/']}>
					<App initialTab={3} />
				</Route>
			</MemoryRouter>
		)

		await waitFor(() => comp.getByTestId('app'))
	})

	it('matches snapshot', () => {
		expect(comp.asFragment()).toMatchSnapshot()
	})
})
