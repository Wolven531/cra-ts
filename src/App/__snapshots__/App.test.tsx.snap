// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches loading snapshot 1`] = `
<DocumentFragment>
  <article
    class="app"
    data-testid="app"
  >
    <header>
      <h3>
        TS Playground
      </h3>
    </header>
    <section
      class="tabs"
    >
      <ul>
        <li
          class="selected"
        >
          tab 1
        </li>
        <li
          class=""
        >
          tab 2
        </li>
        <li
          class=""
        >
          tab 3
        </li>
      </ul>
    </section>
    <section
      class="content"
      data-testid="content"
    >
      <div>
        Content 1
      </div>
    </section>
    <footer>
      © 2021 Anthony Williams
    </footer>
  </article>
</DocumentFragment>
`;

exports[`App w/ initialTab=2 when loaded matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="app"
    data-testid="app"
  >
    <header>
      <h3>
        TS Playground
      </h3>
    </header>
    <section
      class="tabs"
    >
      <ul>
        <li
          class=""
        >
          tab 1
        </li>
        <li
          class="selected"
        >
          tab 2
        </li>
        <li
          class=""
        >
          tab 3
        </li>
      </ul>
    </section>
    <section
      class="content"
      data-testid="content"
    >
      <div>
        Content 2
      </div>
    </section>
    <footer>
      © 2021 Anthony Williams
    </footer>
  </article>
</DocumentFragment>
`;

exports[`App w/ initialTab=3 when loaded matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="app"
    data-testid="app"
  >
    <header>
      <h3>
        TS Playground
      </h3>
    </header>
    <section
      class="tabs"
    >
      <ul>
        <li
          class=""
        >
          tab 1
        </li>
        <li
          class=""
        >
          tab 2
        </li>
        <li
          class="selected"
        >
          tab 3
        </li>
      </ul>
    </section>
    <section
      class="content"
      data-testid="content"
    >
      <div>
        Content 3
      </div>
    </section>
    <footer>
      © 2021 Anthony Williams
    </footer>
  </article>
</DocumentFragment>
`;

exports[`App when loaded matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="app"
    data-testid="app"
  >
    <header>
      <h3>
        TS Playground
      </h3>
    </header>
    <section
      class="tabs"
    >
      <ul>
        <li
          class="selected"
        >
          tab 1
        </li>
        <li
          class=""
        >
          tab 2
        </li>
        <li
          class=""
        >
          tab 3
        </li>
      </ul>
    </section>
    <section
      class="content"
      data-testid="content"
    >
      <div>
        Content 1
      </div>
    </section>
    <footer>
      © 2021 Anthony Williams
    </footer>
  </article>
</DocumentFragment>
`;
